File: Exercicios/Labs/LaboratorioHash.c
#include <stdio.h>

#define M 11        // area principal
#define R 3         // area de reserva
#define TAM 14      // M + R
#define MAX_NOME 100

//tabela hash com area de reserva
typedef struct {
    char tabela[TAM][MAX_NOME]; // paises
    int ocupado[TAM];           // marca se a posicao esta ocupada
    int nr;                     // qnt de elementos na reserva
} TabelaHash;

// compara strings
int comparar(char a[], char b[]) {
    int i = 0;

    while (a[i] != '\0' && b[i] != '\0') {
        if (a[i] != b[i]) {
            return 0;
        }

        i++;
    }

    return a[i] == '\0' && b[i] == '\0';
}

// copia uma string para outra 
void copiar(char destino[], char origem[]) {
    int i = 0;

    while (origem[i] != '\0') {
        destino[i] = origem[i];
        i++;
    }

    destino[i] = '\0';
}

// calcula a funcao hash somando os codigos dos caracteres
int hash(char chave[]) {
    int soma = 0;

    for (int i = 0; chave[i] != '\0'; i++) {
        soma += chave[i];
    }

    return soma % M;
}

// inicializa a tabela hash
void iniciarTabela(TabelaHash *t) {
    t->nr = 0;

    for (int i = 0; i < TAM; i++) {
        t->ocupado[i] = 0;
        t->tabela[i][0] = '\0';
    }
}

// pesquisa um pais na tabela
int pesquisar(TabelaHash *t, char chave[]) {
    int pos = hash(chave);

    // verifica primeiro a posicao principal
    if (t->ocupado[pos] && comparar(t->tabela[pos], chave)) {
        return pos;
    }

    // procura na area de reserva
    for (int i = M; i < M + t->nr; i++) {
        if (t->ocupado[i] && comparar(t->tabela[i], chave)) {
            return i;
        }
    }

    return -1;
}

// insere um pais 
int inserir(TabelaHash *t, char chave[]) {
    // Nao permite inserir duplicado
    if (pesquisar(t, chave) != -1) {
        return 0;
    }

    int pos = hash(chave);

    if (!t->ocupado[pos]) {
        copiar(t->tabela[pos], chave);
        t->ocupado[pos] = 1;
        return 1;
    }

    // se houve colisao, tenta inserir na reserva
    if (t->nr < R) {
        int posReserva = M + t->nr;

        copiar(t->tabela[posReserva], chave);
        t->ocupado[posReserva] = 1;
        t->nr++;

        return 1;
    }

    // se a reserva estiver cheia, nao insere
    return 0;
}

// procura na reserva algum elemento que pertence ao mesmo hash
int procurarNaReservaMesmoHash(TabelaHash *t, int posHash) {
    for (int i = M; i < M + t->nr; i++) {
        if (t->ocupado[i] && hash(t->tabela[i]) == posHash) {
            return i;
        }
    }

    return -1;
}

// remove uma posicao da reserva e reorganiza os elementos
void removerDaReserva(TabelaHash *t, int pos) {
    for (int i = pos; i < M + t->nr - 1; i++) {
        copiar(t->tabela[i], t->tabela[i + 1]);
        t->ocupado[i] = t->ocupado[i + 1];
    }

    int ultima = M + t->nr - 1;

    t->tabela[ultima][0] = '\0';
    t->ocupado[ultima] = 0;
    t->nr--;
}

// remove um pais
int remover(TabelaHash *t, char chave[]) {
    int pos = hash(chave);

    if (t->ocupado[pos] && comparar(t->tabela[pos], chave)) {
        t->tabela[pos][0] = '\0';
        t->ocupado[pos] = 0;

        // verifica se existe na reserva alguem que deve ocupar essa posicao
        int posReserva = procurarNaReservaMesmoHash(t, pos);

        if (posReserva != -1) {
            copiar(t->tabela[pos], t->tabela[posReserva]);
            t->ocupado[pos] = 1;

            removerDaReserva(t, posReserva);
        }

        return 1;
    }

    // caso o pais esteja na area de reserva
    for (int i = M; i < M + t->nr; i++) {
        if (t->ocupado[i] && comparar(t->tabela[i], chave)) {
            removerDaReserva(t, i);
            return 1;
        }
    }

    return 0;
}

// mostra a tabela 
void mostrar(TabelaHash *t) {
    for (int i = 0; i < TAM; i++) {
        if (i < M) {
            printf("%d Principal ", i);
        } else {
            printf("%d Reserva ", i);
        }

        if (t->ocupado[i]) {
            printf("%s\n", t->tabela[i]);
        } else {
            printf("-\n");
        }
    }
}

int main() {
    TabelaHash tabela;
    iniciarTabela(&tabela);

    char comando;
    char pais[MAX_NOME];

    // ate o fim do arquivo
    while (scanf(" %c", &comando) != EOF) {
        if (comando == 'I') {
            scanf("%s", pais);

            if (!inserir(&tabela, pais)) {
                printf("Erro\n");
            }

        } else if (comando == 'P') {
            scanf("%s", pais);

            if (pesquisar(&tabela, pais) != -1) {
                printf("Sim\n");
            } else {
                printf("Nao\n");
            }

        } else if (comando == 'R') {
            scanf("%s", pais);

            if (!remover(&tabela, pais)) {
                printf("Erro\n");
            }

        } else if (comando == 'M') {
            mostrar(&tabela);
        }
    }

    return 0;
}

File: Exercicios/Labs/LaboratorioHash.java
class TabelaHashComReserva {
    private String[] tabela;
    private int m;
    private int r;
    private int nr;

    public TabelaHashComReserva(int m, int r) {
        this.m = m;
        this.r = r;
        this.nr = 0;
        this.tabela = new String[m + r];
    }

    private int hash(String chave) {
        int soma = 0;

        for (int i = 0; i < chave.length(); i++) {
            soma += chave.charAt(i);
        }

        return soma % m;
    }

    private boolean isPosicaoLivre(int pos) {
        return tabela[pos] == null;
    }

    public void inserir(String chave) throws Exception {
        if (pesquisar(chave) != null) {
            throw new Exception("Erro: insercao duplicada");
        }

        int pos = hash(chave);

        if (isPosicaoLivre(pos)) {
            tabela[pos] = chave;
        } else if (nr < r) {
            tabela[m + nr] = chave;
            nr++;
        } else {
            throw new Exception("Erro: area de reserva cheia");
        }
    }

    public String pesquisar(String chave) {
        int pos = hash(chave);

        if (tabela[pos] != null && tabela[pos].equals(chave)) {
            return tabela[pos];
        }

        for (int i = m; i < m + nr; i++) {
            if (tabela[i] != null && tabela[i].equals(chave)) {
                return tabela[i];
            }
        }

        return null;
    }

    public String remover(String chave) {
        int pos = hash(chave);

        if (tabela[pos] != null && tabela[pos].equals(chave)) {
            String removido = tabela[pos];
            tabela[pos] = null;

            int posReserva = procurarNaReservaMesmoHash(pos);

            if (posReserva != -1) {
                tabela[pos] = tabela[posReserva];
                removerDaReserva(posReserva);
            }

            return removido;
        }

        for (int i = m; i < m + nr; i++) {
            if (tabela[i] != null && tabela[i].equals(chave)) {
                String removido = tabela[i];
                removerDaReserva(i);
                return removido;
            }
        }

        return null;
    }

    private int procurarNaReservaMesmoHash(int posHash) {
        for (int i = m; i < m + nr; i++) {
            if (tabela[i] != null && hash(tabela[i]) == posHash) {
                return i;
            }
        }

        return -1;
    }

    private void removerDaReserva(int pos) {
        for (int i = pos; i < m + nr - 1; i++) {
            tabela[i] = tabela[i + 1];
        }

        tabela[m + nr - 1] = null;
        nr--;
    }

    public void mostrar() {
        System.out.println("Indice Area Conteudo");

        for (int i = 0; i < tabela.length; i++) {
            String area;

            if (i < m) {
                area = "Principal";
            } else {
                area = "Reserva";
            }

            String conteudo;

            if (tabela[i] == null) {
                conteudo = "-";
            } else {
                conteudo = tabela[i];
            }

            System.out.println(i + " " + area + " " + conteudo);
        }
    }
}

class TabelaHashComRehash {
    private String[] tabela;
    private int m;

    public TabelaHashComRehash(int m) {
        this.m = m;
        this.tabela = new String[m];
    }

    private int hash(String chave) {
        int soma = 0;

        for (int i = 0; i < chave.length(); i++) {
            soma += chave.charAt(i);
        }

        return soma % m;
    }

    private int rehash(String chave) {
        int soma = 0;

        for (int i = 0; i < chave.length(); i++) {
            soma += chave.charAt(i);
        }

        return (soma + 1) % m;
    }

    public void inserir(String chave) throws Exception {
        if (pesquisar(chave) != null) {
            throw new Exception("Erro: insercao duplicada");
        }

        int pos = hash(chave);

        if (tabela[pos] == null) {
            tabela[pos] = chave;
        } else {
            pos = rehash(chave);

            if (tabela[pos] == null) {
                tabela[pos] = chave;
            } else {
                throw new Exception("Erro: nao foi possivel inserir");
            }
        }
    }

    public String pesquisar(String chave) {
        int pos = hash(chave);

        if (tabela[pos] != null && tabela[pos].equals(chave)) {
            return tabela[pos];
        }

        pos = rehash(chave);

        if (tabela[pos] != null && tabela[pos].equals(chave)) {
            return tabela[pos];
        }

        return null;
    }

    public String remover(String chave) {
        int pos = hash(chave);

        if (tabela[pos] != null && tabela[pos].equals(chave)) {
            String removido = tabela[pos];
            tabela[pos] = null;
            return removido;
        }

        pos = rehash(chave);

        if (tabela[pos] != null && tabela[pos].equals(chave)) {
            String removido = tabela[pos];
            tabela[pos] = null;
            return removido;
        }

        return null;
    }

    public void mostrar() {
        System.out.println("Indice Conteudo");

        for (int i = 0; i < tabela.length; i++) {
            String conteudo;

            if (tabela[i] == null) {
                conteudo = "-";
            } else {
                conteudo = tabela[i];
            }

            System.out.println(i + " " + conteudo);
        }
    }
}

class Celula {
    String elemento;
    Celula prox;

    Celula(String elemento) {
        this.elemento = elemento;
        this.prox = null;
    }
}

class TabelaHashEncadeada {
    private Celula[] tabela;
    private int m;

    public TabelaHashEncadeada(int m) {
        this.m = m;
        this.tabela = new Celula[m];
    }

    private int hash(String chave) {
        int soma = 0;

        for (int i = 0; i < chave.length(); i++) {
            soma += chave.charAt(i);
        }

        return soma % m;
    }

    public void inserir(String chave) throws Exception {
        if (pesquisar(chave) != null) {
            throw new Exception("Erro: insercao duplicada");
        }

        int pos = hash(chave);

        Celula nova = new Celula(chave);
        nova.prox = tabela[pos];
        tabela[pos] = nova;
    }

    public String pesquisar(String chave) {
        int pos = hash(chave);

        Celula atual = tabela[pos];

        while (atual != null) {
            if (atual.elemento.equals(chave)) {
                return atual.elemento;
            }

            atual = atual.prox;
        }

        return null;
    }

    public String remover(String chave) {
        int pos = hash(chave);

        Celula atual = tabela[pos];
        Celula anterior = null;

        while (atual != null) {
            if (atual.elemento.equals(chave)) {
                if (anterior == null) {
                    tabela[pos] = atual.prox;
                } else {
                    anterior.prox = atual.prox;
                }

                return atual.elemento;
            }

            anterior = atual;
            atual = atual.prox;
        }

        return null;
    }

    public void mostrar() {
        System.out.println("Indice Conteudo");

        for (int i = 0; i < tabela.length; i++) {
            System.out.print(i + " ");

            Celula atual = tabela[i];

            if (atual == null) {
                System.out.print("-");
            } else {
                while (atual != null) {
                    System.out.print(atual.elemento);

                    if (atual.prox != null) {
                        System.out.print(" -> ");
                    }

                    atual = atual.prox;
                }
            }

            System.out.println();
        }
    }
}

public class LaboratorioHash {
    public static void main(String[] args) {
        String[] paises = {
            "Brasil", "Canada", "Franca", "China", "Egito",
            "Australia", "Alemanha", "Japao", "Mexico", "Marrocos"
        };

        System.out.println("=== PARTE 1 - HASH COM AREA DE RESERVA ===");

        TabelaHashComReserva reserva = new TabelaHashComReserva(11, 3);

        try {
            for (int i = 0; i < paises.length; i++) {
                reserva.inserir(paises[i]);
            }

            reserva.mostrar();

            System.out.println();
            System.out.println("Pesquisando paises inseridos:");

            for (int i = 0; i < paises.length; i++) {
                if (reserva.pesquisar(paises[i]) != null) {
                    System.out.println(paises[i] + " encontrado");
                } else {
                    System.out.println(paises[i] + " nao encontrado");
                }
            }

            System.out.println();
            System.out.println("Pesquisando paises nao inseridos:");

            if (reserva.pesquisar("Estados Unidos") == null) {
                System.out.println("Estados Unidos nao encontrado");
            }

            if (reserva.pesquisar("Argentina") == null) {
                System.out.println("Argentina nao encontrado");
            }

            System.out.println();
            System.out.println("Testando insercao duplicada:");

            try {
                reserva.inserir("Brasil");
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }

            System.out.println();
            System.out.println("Testando overflow:");

            try {
                reserva.inserir("Espanha");
            } catch (Exception e) {
                System.out.println(e.getMessage());
            }

            System.out.println();
            System.out.println("Removendo Brasil, China e Alemanha:");

            reserva.remover("Brasil");
            reserva.remover("China");
            reserva.remover("Alemanha");

            reserva.mostrar();

        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        System.out.println();
        System.out.println("=== PARTE 2 - HASH COM REHASH ===");

        TabelaHashComRehash rehash = new TabelaHashComRehash(11);

        try {
            for (int i = 0; i < paises.length; i++) {
                rehash.inserir(paises[i]);
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        rehash.mostrar();

        System.out.println();
        System.out.println("=== PARTE 3 - HASH COM LISTA ENCADEADA ===");

        TabelaHashEncadeada encadeada = new TabelaHashEncadeada(11);

        try {
            for (int i = 0; i < paises.length; i++) {
                encadeada.inserir(paises[i]);
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

        encadeada.mostrar();
    }
}